File: rxbinding-leanback-v17/src/main/java/com/jakewharton/rxbinding2/support/v17/leanback/widget/SearchBarSearchQueryChangesOnSubscribe.java
package com.jakewharton.rxbinding2.support.v17.leanback.widget;

import android.support.v17.leanback.widget.SearchBar;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static io.reactivex.android.MainThreadDisposable.verifyMainThread;

final class SearchBarSearchQueryChangesOnSubscribe extends Observable<String> {
  final SearchBar view;

  SearchBarSearchQueryChangesOnSubscribe(SearchBar view) {
    this.view = view;
  }

  @Override protected void subscribeActual(final Observer<? super String> observer) {
    verifyMainThread();
    Listener listener = new Listener(view, observer);
    observer.onSubscribe(listener);
    view.setSearchBarListener(listener);
  }

  static class Listener extends MainThreadDisposable implements SearchBar.SearchBarListener {
    final SearchBar view;
    final Observer<? super String> observer;

    Listener(SearchBar view, Observer<? super String> observer) {
      this.view = view;
      this.observer = observer;
    }

    @Override public void onSearchQueryChange(String query) {
      if (!isDisposed()) {
        observer.onNext(query);
      }
    }

    @Override public void onSearchQuerySubmit(String query) {
    }

    @Override public void onKeyboardDismiss(String query) {
    }

    @Override protected void onDispose() {
      view.setSearchBarListener(null);
    }
  }
}

File: rxbinding-leanback/src/main/java/com/jakewharton/rxbinding3/leanback/SearchBarSearchQueryChangesOnSubscribe.kt
@file:JvmName("RxSearchBar")
@file:JvmMultifileClass

package com.jakewharton.rxbinding3.leanback

import androidx.annotation.CheckResult
import androidx.leanback.widget.SearchBar
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.android.MainThreadDisposable

import com.jakewharton.rxbinding2.internal.Preconditions.checkMainThread

/**
 * Create an observable of String values for search query changes on `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 */
@CheckResult
fun SearchBar.searchQueryChanges(): Observable<String> {
  return SearchBarSearchQueryChangesOnSubscribe(this)
}

private class SearchBarSearchQueryChangesOnSubscribe(
  private val view: SearchBar
) : Observable<String>() {

  override fun subscribeActual(observer: Observer<in String>) {
    if (!checkMainThread(observer)) {
      return
    }
    val listener = Listener(view, observer)
    observer.onSubscribe(listener)
    view.setSearchBarListener(listener)
  }

  private class Listener(
    private val view: SearchBar,
    private val observer: Observer<in String>
  ) : MainThreadDisposable(), SearchBar.SearchBarListener {

    override fun onSearchQueryChange(query: String) {
      if (!isDisposed) {
        observer.onNext(query)
      }
    }

    override fun onSearchQuerySubmit(query: String) {}

    override fun onKeyboardDismiss(query: String) {}

    override fun onDispose() {
      view.setSearchBarListener(null)
    }
  }
}
